QWT-cross sync

jevank opened the pull request:
Apply qubes-windows-tools-cross patches. See commit messages to details

Commit:
Bugfix: advertise-tools issues error when calls qubes.NotifyTools in dom0 due to non-existing dummy process

qrexec-wrapper.exe now acts as local process if local process is "(null)"

Commit:
qrexec-agent: call libvchan_close when service is stopped

Commit:
qrexec-agent, qrexec-client-vm: remove duplicated error logs; add more detailed error messages

Commit:
qrexec-wrapper: XifLogger: do not decrease log level

jevank committed:
qrexec-agent: fix relative path construct

This fixes qubes-rpc service fail with argument appended.

jevank committed:
qrexec: do not break with protocol mismatch

This requires to support R4.1 qrexec v3

jevank committed:
Add support for MinGW build

- make all includes lowercase (Windows don't care, MinGW on Linux ships
  them lowecase)
- add few workarounds for missing headers/features in MinGW
- where in doubt, enclouse changes in __MINGW32__ to keep compatibility
  with original VS build
- rename perror to win_perror
- add Makefile for MinGW build

jevank committed:
build with inlined __chkstk_ms

jevank committed:
relocate-dir: support IO_REPARSE_TAG_APPEXECLINK

This fixes move-profile feature fails on Windows 10

This fixes QubesOS/qubes-issues#5093

jevank committed:
qrexec-service: wait for process to complete

This fixes dispvm halts right after process was invoked

Commit:
bugfix: HandleExec: send MSG_CONNECTION_TERMINATED when process is terminated

marmarek commented on src/qrexec-agent/qrexec-agent.c:
```diff
@@ -926,9 +927,9 @@ static DWORD HandleDaemonHello(struct msg_header *header)
+        LogWarning("incompatible protocol version (%d instead of %d)",
                  info.version, QREXEC_PROTOCOL_VERSION);
-        return ERROR_INVALID_FUNCTION;
+        //return ERROR_INVALID_FUNCTION;
```
I'd prefer to change the condition to `info.version < QREXEC_PROTOCOL_VERSION` instead. The protocol got negotiation support, which works like this:
1. Both sides send their maximum supported version.
2. Both sides look at what other side sent them.
3. Choose minimum.
4. If that version is supported, continue, otherwise close the connection.

It means, the windows agent will be fine with sending version `2`, if the daemon sends at least `2` too.

When adding support for version `3`, it will need to either save what version was negotiated, or drop support for version `2`. But that's for another PR.

marmarek commented on src/qrexec-wrapper/qrexec-wrapper.c:
```diff
@@ -552,7 +554,7 @@ DWORD HandleDataMessage(
         {
             LogWarning("incompatible protocol version (got %d, expected %d)", peerInfo.version, QREXEC_PROTOCOL_VERSION);
-            return ERROR_INVALID_FUNCTION;
+            //return ERROR_INVALID_FUNCTION;
```
Same comment as the other place.

Commit:
qrexec: check minimum supported protocol version

jevank commented:
removed last commit with rexec connection termination support, it is not ready yet

Commit:
Add qubes.SuspendPostAll with time sync

This fixes the desynchronization of time after sleep/resume actions

Commit:
clipboard-copy: wipe \r symbols which break paste

michaellukashov committed:
filecopy: improve filename convertion

Replace illegal filename symbols by '_'

copied from git: xutftowcs_path

This fixes QubesOS/qubes-issues#5819

marmarek commented on src/qrexec-agent/qrexec-agent.c:
```diff
@@ -70,6 +70,95 @@ void DumpRequestList(void)
+    {
+        if (connection_info[i].handle == NULL)
+        {
+            EnterCriticalSection(&g_DaemonCriticalSection);
```
The whole loop should be under the lock. Otherwise two threads could write to the same slot.

marmarek commented on src/qrexec-agent/qrexec-agent.c:
```diff
@@ -70,6 +70,95 @@ void DumpRequestList(void)
+        {
+            if (!handle_is_running(handle))
+            {
+                EnterCriticalSection(&g_DaemonCriticalSection);
```
At the very least reading `connection_info[i].handle` should be already under the lock - otherwise another thread could release the slot in the meantime (resulting in releasing the same slot twice, at best). But it's safer to put the whole loop under the lock already.

marmarek commented on src/qrexec-agent/qrexec-agent.c:
```diff
@@ -1075,9 +1171,10 @@ static DWORD WatchForEvents(HANDLE stopEvent)
+                win_perror2(status, "HandleDaemonMessage");
                 goto out;
             }
+            check_connections();
```
Checking it here results in a funny effect - that port is released only after subsequent connection is made. So, repeated connections use port 513 (first available), then 514 (because 513 wasn't released _yet_) and then again 513 and so on.
For this case it isn't a big deal, but if you'd have windows-based DispVM, it would be destroyed with a delay (until subsequent qrexec call). So, a better solution is to include those handles in `WaitForMultipleObjects` call.

jevank replied:
Do you mean to extend waitObjects array with these handlers?

marmarek replied:
yes

jevank replied:
fixed

marmarek commented on src/qrexec-services/clipboard-copy/clipboard-copy.c:
```diff
@@ -30,56 +30,90 @@
+    if (!clipText || !clipTextPrepared)
+        return FALSE;
+    // replace \r\n -> \r
+    *clipTextPrepared = (WCHAR*)calloc((wcslen(clipText) + 1) * sizeof(WCHAR), 1);
```
missing error check

marmarek commented on src/qrexec-services/file-receiver/unpack.c:
```diff
@@ -104,8 +194,9 @@ void ProcessRegularFile(IN const struct file_header *untrustedHeader, IN const c
 
-    errorCode = ConvertUTF8ToUTF16(untrustedNameUtf8, &untrustedFileName, NULL);
-    if (ERROR_SUCCESS != errorCode)
+    untrustedFileName = (WCHAR*)calloc(MAX_PATH + 1, 1);
```
missing error check

marmarek commented on src/qrexec-services/file-receiver/unpack.c:
```diff
@@ -155,8 +246,9 @@ void ProcessDirectory(IN const struct file_header *untrustedHeader, IN const cha
 
-    errorCode = ConvertUTF8ToUTF16(untrustedNameUtf8, &untrustedDirectoryName, NULL);
-    if (ERROR_SUCCESS != errorCode)
+    untrustedDirectoryName = (WCHAR*)calloc(MAX_PATH + 1, 1);
```
missing error check

marmarek commented on src/qrexec-services/file-receiver/unpack.c:
```diff
@@ -189,11 +281,11 @@ void ProcessLink(IN const struct file_header *untrustedHeader, IN const char *un
 
-    errorCode = ConvertUTF8ToUTF16(untrustedNameUtf8, &untrustedName, NULL);
-    if (ERROR_SUCCESS != errorCode)
+    untrustedName = (WCHAR*)calloc(MAX_PATH + 1, 1);
```
missing error check

marmarek commented on src/qrexec-services/file-receiver/unpack.c:
```diff
@@ -218,8 +310,9 @@ void ProcessLink(IN const struct file_header *untrustedHeader, IN const char *un
 
-    errorCode = ConvertUTF8ToUTF16(untrustedLinkTargetPathUtf8, &untrustedLinkTargetPath, NULL);
-    if (ERROR_SUCCESS != errorCode)
+    untrustedLinkTargetPath = (WCHAR*)calloc(MAX_PATH + 1, 1);
```
missing error check

Commit:
qrexec-agent: check registered processes in main loop

Commit:
clipboard-copy: add calloc checks

Commit:
file-receiver: add calloc checks
